### PR TITLE
Exclude NFA start state from its own epsilon closure for reversed FSM with multiple end states

### PR DESCRIPTION
**Pull request by katef**
This addresses the situation described by #62.

### CODE REVIEW

File: include/fsm/pred.h [hunk 1]
```diff
@@ -37,5 +37,11 @@ fsm_hasincoming(const struct fsm *fsm, const struct fsm_state *state);
 int
 fsm_hasoutgoing(const struct fsm *fsm, const struct fsm_state *state);
 
+/*
+ * True iff there are outgoing edges and all are epsilons.
+ */
+int
+fsm_epsilonsonly(const struct fsm *fsm, const struct fsm_state *state);
```
Comment (sfstewman) [inline include/fsm/pred.h]: I like the predicate approach for this.

File: src/libfsm/determinise.c [hunk 2]
```diff
@@ -218,6 +219,15 @@ state_closure(struct set *mappings, struct fsm *dfa, const struct fsm_state *nfa
...
+
+		if (set_count(ec) == 0) {
+			set_free(ec);
+			ec = NULL;
+		}
+	}
+
 	if (ec == NULL) {
 		return NULL;
 	}
```
Comment (sfstewman) [inline src/libfsm/determinise.c]: So, given that you call state_closure a lot, and only remove a state once (at the beginning) under very special conditions, does it make sense to pass this as a boolean into the state_closure function?  I'd be tempted to either write a special remove function or to inline the removal.

File: src/libfsm/determinise.c [hunk 3]
```diff
@@ -450,9 +460,34 @@ determinise(struct fsm *nfa,
...
+		 * This is the situation we detect here.
+		 *
+		 * This fabricated start state is then excluded from its epsilon closure,
+		 * so that the closures for its destination states are found equivalent,
+		 * because they also do not include the start state.
+		 *
+		 * If you pass an equivalent NFA where this is not the case (for example,
+		 * with the start state containing an epsilon edge to itself), we regard
+		 * this as any other DFA, and minimal output is not guaranteed.
+		 */
```
Comment (sfstewman) [inline src/libfsm/determinise.c]: The comment looks good to me.  Possibly worth including a reference to the StackExchange answer https://cs.stackexchange.com/questions/6138/problem-with-implementing-brzozowskis-algorithm?noredirect=1&lq=1

File: src/libfsm/determinise.c (hunk 2 above)
Comment (katef) [reply src/libfsm/determinise.c]: I did wonder that. There's be no other consumer for `state_closure()` with its own state included, so I'd just delete it. And then I got unhappy about losing information. So not today - perhaps when dredging some of this stuff out to a more general interface.

File: src/libfsm/determinise.c (hunk 3 above)
Comment (katef) [reply src/libfsm/determinise.c]: I saw that! The answer was too terse for me to follow. So I don't think it'd suit this audience, either.

**Pull request merged**